### PR TITLE
fix(emoji-text-wrapper): DLT-1699 emoji alignment

### PR DESCRIPTION
**Pull request by juliodialpad**
# Emoji text wrapper - Fix emoji alignment

## Obligatory GIF (super important!)

Obligatory GIF

## :hammer_and_wrench: Type Of Change

These types will increment the version number on release:

- [x] Fix

## :book: Jira Ticket

https://dialpad.atlassian.net/browse/DLT-1699

## :book: Description

- Changed the output of emoji text-wrapper, text-only nodes are retained as-is and added a wrapper container to be able to vertically center and truncate text
- Fixed an issue that prevented the emoji size to change on storybook.

## :bulb: Context

Emoji on leftbar row was not aligned correctly but it was using emoji-text-wrapper so fixed directly on the source.

## :pencil: Checklist

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [x] I have reviewed my changes.
- [ ] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

For all Vue changes:

- [ ] I have added / updated unit tests.
- [x] I have made my changes in Vue 2 and Vue 3. Note: you may sync your changes from Vue 2 to Vue 3 (or vice versa) using the `./scripts/dialtone-vue-sync.sh` script. Read docs here: [Dialtone Vue Sync Script](../packages/dialtone-vue3/.github/CONTRIBUTING.md#dialtone-vue-sync-script)
- [ ] I have validated components with a screen reader.
- [ ] I have validated components keyboard navigation.

For all CSS changes:

- [ ] I have used design tokens whenever possible.
- [x] I have considered how this change will behave on different screen sizes.
- [ ] I have visually validated my change in light and dark mode.
- [x] I have used gap or flexbox properties for layout instead of margin whenever possible.

## :camera: Screenshots / GIFs

- Before
image

- After
Screenshot 2024-07-19 at 3 22 50 p m

## Shortcuts to affected components
https://dialtone.dialpad.com/vue/deploy-previews/pr-407/?path=/story/components-emoji-text-wrapper--default
https://dialtone.dialpad.com/vue/deploy-previews/pr-407/?path=/story/components-emoji-text-wrapper--variants
https://dialtone.dialpad.com/vue/deploy-previews/pr-407/?path=/story/recipes-conversation-view-emoji-row--default
https://dialtone.dialpad.com/vue/deploy-previews/pr-407/?path=/story/recipes-conversation-view-feed-item-row--default
https://dialtone.dialpad.com/vue/deploy-previews/pr-407/?path=/story/recipes-leftbar-contact-row--variants
https://dialtone.dialpad.com/vue/deploy-previews/pr-407/?path=/story/recipes-leftbar-general-row--variants

### CODE REVIEW

**Comment by juliodialpad** [comment]
This is finally ready for review, visual changes are expected as I changed the vertical alignment of the text and the emojis so they are completely centered.

@francisrupert @fede-dp

File: packages/dialtone-css/lib/build/less/components/emoji.less
Comment (braddialpad) [inline packages/dialtone-css/lib/build/less/components/emoji.less]: hmm.. slightly worried this could break stuff?

File: packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less
Comment (braddialpad) [inline packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less]: absolutely crazy this is what we have to do to center 😆

File: packages/dialtone-vue2/recipes/conversation_view/emoji_row/emoji_row.vue
Comment (braddialpad) [inline packages/dialtone-vue2/recipes/conversation_view/emoji_row/emoji_row.vue]: Shouldn't it already be this from your change in emoji.less?

File: packages/dialtone-css/lib/build/less/components/emoji.less
Comment (juliodialpad) [reply packages/dialtone-css/lib/build/less/components/emoji.less]: I think this change is not necessary, reverting this one, I just need to change this on emoji-text-wrapper in case the display property needs to change.

File: packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less
Comment (juliodialpad) [reply packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less]: Yeah, centering a changing height image with a text it's more complex that it seemed at first sight. But I think this is the safest way to do it even if the text size change.

File: packages/dialtone-vue2/recipes/conversation_view/emoji_row/emoji_row.vue
Comment (juliodialpad) [reply packages/dialtone-vue2/recipes/conversation_view/emoji_row/emoji_row.vue]: This is not needed either, was a leftover of my multiple tries

**Comment by braddialpad** [comment]
Supposed to be this tight to the edge?

Screenshot 2024-07-22 at 10 24 47 AM

here is what it looked like prior:

Screenshot 2024-07-22 at 10 24 51 AM

**Comment by juliodialpad** [comment]
> Supposed to be this tight to the edge?
> 
> Screenshot 2024-07-22 at 10 24 47 AM
> 
> here is what it looked like prior:
> 
> Screenshot 2024-07-22 at 10 24 51 AM

Just fixed, it needed the `display: inline-flex` on emoji-text-wrapper 😄 should be updated now.

**Review by braddialpad — APPROVED** [review]

**Comment by juliodialpad** [comment]
I would like to confirm with @francisrupert that this is the result that we're expecting before merging. Please approve on Percy if this is ok

File: packages/dialtone-vue2/components/emoji_text_wrapper/emoji_text_wrapper.vue
Comment (iropolo) [inline packages/dialtone-vue2/components/emoji_text_wrapper/emoji_text_wrapper.vue]: huh, we always pass an array, I guess we could remove the optional chaining operator

**Review by iropolo — APPROVED** [review]

File: packages/dialtone-vue2/components/emoji_text_wrapper/emoji_text_wrapper.vue
Comment (juliodialpad) [reply packages/dialtone-vue2/components/emoji_text_wrapper/emoji_text_wrapper.vue]: Sure, rememeber adding that due to an error I got but I fixed that code so it's not needed right now.

**Comment by francisrupert** [comment]
Short answer for now: there's a better way to ensure the baseline doesn't shift while guaranteeing the Emoji is consistently placed next to text regardless of emoji size. Additionally, I really don't like how Emoji grows the line-height. 

I'll play with something. I've done this before, just gotta make it work for the tag soup Emoji Wrapper introduces.

File: packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less:14]: @juliodialpad, 

OK, here's what I've landed on: https://codepen.io/francisrupert/pen/OJeXjGO?editors=1100

This controls the height of the Emoji itself by inheriting, specifically `height: 1em`. It then positions upward using the age-old `top: 50%; transform: translateY(-50%);`.

So yes, bigger Emoji sizes in smaller text will bleed out... though that will probably only occur if someone **decides** the emoji should be a bigger size and accepts the overlap.

Short screencast walkthrough attached. 

https://github.com/user-attachments/assets/206cc561-e15f-41f9-80f5-a3aa940b13fa

**Review by francisrupert — CHANGES_REQUESTED** [review]

File: packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less
Comment (francisrupert) [reply packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less:14]: Note that I renamed the selector with `.ZZZ-` just to ensure it doesn't collide with any existing `d-emoji-text-wrapper`. Just rename it back and adjust as needed.

Comment (juliodialpad) [reply packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less:14]: This is the issue I've been having, and this is why I took several days trying to center this and end up with such a "messy" or "dirty" solution, was the only way I found to actually center the emojis in every case.

Screenshot 2024-07-23 at 12 28 30 p m

I'm seeing the emojis are not actually centered if put lines on the text, you can see the emojis are actually lower than expected.

Comment (francisrupert) [reply packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less:14]: Let's do this...

```
top: calc(50% - 1px);
```

Updated on https://codepen.io/francisrupert/pen/OJeXjGO?editors=1100

image

Comment (braddialpad) [reply packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less:14]: > top: calc(50% - 1px);

Will that actually scale correctly with different text sizes and emoji sizes??

Comment (francisrupert) [reply packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less:14]: > ... scale correctly with different text sizes and emoji sizes?

Granted, an absolute nudging value of `1px` may feel odd across all the sizes, it's quite consistent across varying emoji sizes within varying text size and line-height. Can stress-test it at https://codepen.io/francisrupert/pen/OJeXjGO?editors=1100

What I'm most concerned about is shifting baselines and line-height of the parent element or elements.

**Optical** alignment is a finicky thing, so it becomes a design judgement call. While a few of these may not _strictly_ be perfectly centered, their optical alignment is pretty darn good. Additionally, the few that barely aren't perfectly centered aren't likely to occur in practical use — much like bigger emojis won't be used within small text size. 

image

Comment (juliodialpad) [reply packages/dialtone-css/lib/build/less/components/emoji-text-wrapper.less:14]: Alright, I'll do the changes and see how it behaves on visual tests.

**Review by francisrupert — APPROVED** [review]
Yeah, much better and won't negatively shift baselines or introduce added space. 

Related, but not part of this PR is what I called out in [Nina's PR](https://github.com/dialpad/dialtone/pull/412#issuecomment-2246375426). That's why the smaller sizes are out of alignment, even independent of the Emoji Wrapper. 

Fortunately, that won't impact this because the CSS change you added prevents that from being an issue for the Emoji within the Emoji Wrapper because it is positioned. 

image

**Pull request merged**
